subprojects/hashing/src/main/java/org/gradle/internal/hash/FileContentTypeDetectingInputStream.java

```diff
@@ -0,0 +1,117 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.hash;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Infers whether a file is a binary file or not by checking if there are any ASCII
+ * control characters in the file.  If so, then it is likely a binary file.
+ *
+ * Note that all operations should delegate to the underlying input stream, except
+ * for mark() and markSupported() which are unsupported.  The methods newly
+ * introduced in Java 9 (such as readAllBytes(), transferTo(), etc) are left
+ * unimplemented as this class currently must be Java 8 compatible.
+ */
+public class FileContentTypeDetectingInputStream extends InputStream {
+    private final static int CHAR_SCANNING_LIMIT = 80000;
+    private final InputStream delegate;
+    private boolean controlCharactersFound;
+    private int count;
+
+    public FileContentTypeDetectingInputStream(InputStream delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public int read() throws IOException {
+        int next = delegate.read();
+        checkForControlCharacters(next);
+        return next;
+    }
+
+    @Override
+    public int read(byte[] buffer) throws IOException {
+        int read = delegate.read(buffer);
+        checkForControlCharacters(buffer, read);
+        return read;
+    }
+
+    @Override
+    public int read(byte[] buffer, int off, int len) throws IOException {
+        int read = delegate.read(buffer, off, len);
+        checkForControlCharacters(buffer, read);
+        return read;
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
+        return delegate.skip(n);
+    }
+
+    @Override
+    public int available() throws IOException {
+        return delegate.available();
+    }
+
+    @Override
+    public synchronized void reset() throws IOException {
+        count = 0;
+        controlCharactersFound = false;
+        delegate.reset();
+    }
+
+    void checkForControlCharacters(byte[] buffer, int length) {
+        if (count < CHAR_SCANNING_LIMIT && !controlCharactersFound) {
+            for (int i = 0; i < length; i++) {
+                checkForControlCharacters(buffer[i]);
+            }
+        }
+    }
+
+    void checkForControlCharacters(int b) {
+        if (count++ < CHAR_SCANNING_LIMIT && !controlCharactersFound && isControlCharacter(b)) {
+            controlCharactersFound = true;
+        }
+    }
+
+    private boolean isControlCharacter(int c) {
+        return isInControlRange(c) && isNotCommonTextChar(c);
+    }
+
+    private static boolean isInControlRange(int c) {
+        return c >= 0x00 && c < 0x20;
```

Comment:
So all we are trying to do here is apply a heuristic to determine if the file is likely to be a binary file.  The `isISOControl()` method actually adds some additional characters as control characters (i.e. `0x7f` to `0x9f`).  Which is to say that using that method would be _more_ likely to determine that a file is binary.  However, some of those additional characters do show up in UTF-8 encoding, so we would sometimes determine that a file is binary when it is actually text.  By comparison, Git actually leans more in the opposite direction - it only looks for the `0x00` character (see https://git.kernel.org/pub/scm/git/git.git/tree/xdiff-interface.c?h=v2.30.0#n187) rather than the range that we are looking at here.  I could be convinced, but I'm not sure the value of adding more characters here would outweigh the risk of classifying additional text files as binary.